File: src/main/java/org/cloudfoundry/promregator/config/PromregatorConfiguration.java
package org.cloudfoundry.promregator.config;

import java.util.ArrayList;
import java.util.List;

import org.springframework.boot.context.properties.ConfigurationProperties;
import org.springframework.boot.context.properties.EnableConfigurationProperties;
import org.springframework.context.annotation.Configuration;

@ConfigurationProperties(prefix="promregator")
public class PromregatorConfiguration {
	private List<Target> targets = new ArrayList<>();

	private AuthenticatorConfiguration authenticatorConfiguration;
	
	private List<TargetAuthenticatorConfiguration> targetAuthenticators = new ArrayList<>();
	
	public List<Target> getTargets() {
		return targets;
	}

	public void setTargets(List<Target> targets) {
		this.targets = targets;
	}

	public AuthenticatorConfiguration getAuthenticator() {
		return authenticatorConfiguration;
	}

	public void setAuthenticator(AuthenticatorConfiguration authenticatorConfiguration) {
		this.authenticatorConfiguration = authenticatorConfiguration;
	}

	public List<TargetAuthenticatorConfiguration> getTargetAuthenticators() {
		return targetAuthenticators;
	}

	public void setTargetAuthenticators(List<TargetAuthenticatorConfiguration> targetAuthenticators) {
		this.targetAuthenticators = targetAuthenticators;
	}
	
}

File: src/main/java/org/cloudfoundry/promregator/lite/config/PromregatorConfiguration.kt
package org.cloudfoundry.promregator.lite.config

import org.cloudfoundry.promregator.config.AbstractOAuth2XSUAAAuthenticationConfiguration
import org.cloudfoundry.promregator.config.OAuth2XSUAABasicAuthenticationConfiguration
import org.cloudfoundry.promregator.config.OAuth2XSUAACertificateAuthenticationConfiguration
import org.cloudfoundry.promregator.lite.config.PromregatorConfiguration.Companion.DEFAULT_ID
import org.springframework.boot.context.properties.ConfigurationProperties
import org.springframework.boot.context.properties.ConstructorBinding

@ConstructorBinding
@ConfigurationProperties(prefix = "promregator")
data class PromregatorConfiguration(
        val targets: List<Target> = ArrayList(),
        val authenticator: AuthenticatorConfiguration = AuthenticatorConfiguration(),
        val targetAuthenticators: List<TargetAuthenticatorConfiguration> = ArrayList(),
        val internal: InternalConfig = InternalConfig(),
) {
    companion object {
        const val DEFAULT_ID = "__default__"
    }
}

data class InternalConfig(
        val preCheckAPIVersion: Boolean = true,
)

open class AuthenticatorConfiguration(
) {
    var type: String? = null
    var basic: BasicAuthenticationConfiguration? = null
    var oauth2xsuaa: OAuth2XSUAABasicAuthenticationConfiguration? = null
    var oauth2xsuaaBasic: OAuth2XSUAABasicAuthenticationConfiguration? = null
    var oauth2xsuaaCertificate: OAuth2XSUAACertificateAuthenticationConfiguration? = null
}

class BasicAuthenticationConfiguration(
        val username: String,
        val password: String,
)

class OAuth2XSUAAAuthenticationConfiguration {
    var tokenServiceURL: String? = null
    var client_id: String? = null
    var client_secret: String? = null
    var scopes: String? = null
}

class OAuth2XSUAABasicAuthenticationConfiguration : AbstractOAuth2XSUAAAuthenticationConfiguration() {
    var tokenServiceURL: String? = null
    var client_secret: String? = null
}

class OAuth2XSUAACertificateAuthenticationConfiguration : AbstractOAuth2XSUAAAuthenticationConfiguration() {
    var tokenServiceCertURL: String? = null
    var client_certificates: String? = null
    var client_key: String? = null
}

class TargetAuthenticatorConfiguration : AuthenticatorConfiguration() {
    var id: String = DEFAULT_ID
}
